### DISCUSSION TITLE
Mapping non-contiguous values

### DISCUSSION THREAD

**Discussion by jeremywiebe** [jam1garner/binrw#225] [Q&A] [answered]
Pretty early Rust/binrw user here, so I could be missing something. 

I am working to read a binary structure where some of the values are represented in different bytes of the header. 

For example, one u8 value is represented as 4 bits from one byte and 4 bits from another bit.

```
  Byte 1
  87654321
  ||||||||
  ||||++++ (ignored for this question)
  ++++---- bits 0-3 of u8 value

  Byte 2
  87654321
  ||||||||
  ||||++++ (ignored for this question)
  ++++---- bits 4-7 of u8 value
```

So given the following input bytes `0x451a` (binary `01000101 00011010`) the u8 value should be 0x14 (`0001 0100`). 

Currently I just use two `B4` fields and then have a function on the struct to synthesize the value.

**Comment by csnover** [comment, marked as answer]
The only way to handle this case is to synthesise the value like you are doing, but there are other approaches than the approach you are using. You could also take the raw data with `#[br(temp)]` and then split it up into different fields using `#[br(calc)]` so you only calculate the value once at parse time. If the value is not used very often relative to the number of times you are parsing an object with this structure, continuing to use a member function that does it on-demand is probably better, but most of the time this is a micro-optimisation not worth thinking about.

**Reply by jeremywiebe** [reply]
Thanks for confirming. I hadn't noticed `#[br(calc)]` yet, but right now what I have works and perf definitely won't be a problem for what I'm doing. Thanks again.
